draft-ietf-tls-extended-key-update.md

```diff
@@ -584,62 +507,346 @@ The exchange has the following steps:
+                           <-------- [ExtendedKeyUpdate(response)]
+[C: tx=3, rx=3]                   [S: tx=3, rx=3]
 
- [ExtendedKeyUpdate(new_key_update)] -------->
- (epoch=3)
+[C: tx=3, rx=3]                   [S: tx=3, rx=3]
+[ExtendedKeyUpdate(new_key_update)] -------->
+[C: tx=3, rx=3]                   [S: tx=3, rx=3]
 
-        <-------- [ExtendedKeyUpdate(new_key_update)]
-                                          (epoch=3)
+[C: tx=3, rx=3]                   [S: tx=3, rx=3]
+                 <-------- [ExtendedKeyUpdate(new_key_update)]
+[C: tx=3, rx=3]                   [S: tx=3, rx=3]
 
- [ACK]                   -------->
- (epoch=4)
+# Epoch switch point:
+# Client bumps epochs when sending ACK (tx=4, rx=4).
+# Server bumps epochs only *after receiving* that ACK.
 
-                           <--------   [Application Data]
-                                          (epoch=4)
+[C: tx=4, rx=4]                   [S: tx=3, rx=3]
+[ACK] (tag=new, tx==rx==4)        -------->
+[C: tx=4, rx=4]                   [S: tx=4, rx=4]
 
- [Application Data]
- (epoch=4)               -------->
+[C: tx=4, rx=4]                   [S: tx=4, rx=4]
+                                  <--------   [Application Data]
+[C: tx=4, rx=4]                   [S: tx=4, rx=4]
+
+[C: tx=4, rx=4]                   [S: tx=4, rx=4]
+[Application Data]                -------->
+[C: tx=4, rx=4]                   [S: tx=4, rx=4]
 ~~~
 {: #dtls-key-update title="Example DTLS 1.3 Extended Key Update Exchange."}
 
-Due to the possibility of an `ExtendedKeyUpdate(new_key_update)` message being
-lost and thereby preventing the sender of that message from updating its keying
-material, receivers MUST retain the pre-update keying material until receipt
-and successful decryption of a message using the new keys.
+The following table shows the steps, the message in flight, and the tx/rx epochs on both sides.
 
-Due to loss and/or reordering, DTLS 1.3 peers MAY receive a record with an
-older epoch than the current one. They SHOULD attempt to process such records
-for that epoch but MAY opt to discard such out-of-epoch records.
+~~~
++-----+--------------------+-------------+-------+-------------+
+|Step | Message            | Client tx/rx| MsgEp | Server tx/rx|
++-----+--------------------+-------------+-------+-------------+
+|  1  | App Data --------->| 3/3 -> 3/3  |   3   | 3/3 -> 3/3  |
+|  2  | <------ App Data   | 3/3 -> 3/3  |   3   | 3/3 -> 3/3  |
+|  3  | EKU(request) ----->| 3/3 -> 3/3  |   3   | 3/3 -> 3/3  |
+|  4  | <------ EKU(resp)  | 3/3 -> 3/3  |   3   | 3/3 -> 3/3  |
+|  5  | EKU(new_key_upd) ->| 3/3 -> 3/3  |   3   | 3/3 -> 3/3  |
+|  6  | <--- EKU(new_key_u)| 3/3 -> 3/3  |   3   | 3/3 -> 3/3  |
+|  7  | ACK -------------->| 3/3 -> 4/4  |   4   | 3/3 -> 4/4  |
+|  8  | <------ App Data   | 4/4 -> 4/4  |   4   | 4/4 -> 4/4  |
+|  9  | App Data --------->| 4/4 -> 4/4  |   4   | 4/4 -> 4/4  |
++-----+--------------------+-------------+-------+-------------+
```

Comment:
The above table does not align with the following step:

After the initiator receives the responder’s ExtendedKeyUpdate(new_key_update), the initiator MUST update its send key and epoch value. With the receipt of that message, the initiator MUST also update its receive keys.

Reply:
Updated table.